File: 백준/Silver/14501. 퇴사/퇴사.java

import java.util.Scanner;

//알고리즘은 맞았는데 구현에서 살짝 삐끗
public class Main {
    static int n = 0;
    static int[] t = new int[n+1];
    static int[] p = new int[n+1];
    static int max = 0;
    public static void main(String[] args) {



        Scanner sc = new Scanner(System.in);
         n = sc.nextInt();
         t = new int[n+1];
         p = new int[n+1];
        for(int i = 1; i<= n; i++){
             t[i] = sc.nextInt();
             p[i] = sc.nextInt();
        }

      dfs( 1,0);
        System.out.print(max);

    }

    static void dfs( int day, int money){



        if(day == n+1){
            if(max < money)
                max = money;
            return;
        }
        if(day > n+1) return;

            dfs(day+ t[day], money+ p[day]); //해당날짜의 일을 수행하는 경우
            dfs( day+1, money); // 해당날짜의 일을 수행 안하는 경우

    }
}

File: 백준/Silver/14501. 퇴사/퇴사.kt
    import java.util.Scanner

var T = IntArray(0)
    var P = IntArray(0)
    var n = 0
    var max = 0
    fun main(args: Array<String>): Unit = with(Scanner(System.`in`)) {
         n = nextInt()
        T = IntArray(n+1)
        P = IntArray(n+1)

        for (i in 1..n){
            T[i] = nextInt()
            P[i] = nextInt()
        }
        go(1, 0)
        print(max)
    }

    fun go(day:Int, sum:Int){


        if (day == n+1){
            if (max < sum)max = sum
            return
        }
        if (day > n+1) return



            go(day+T[day], sum+P[day])
            go(day+1, sum)

    }